### PR TITLE
Main Processor

### PR DESCRIPTION
**Pull request by FinlayPatoto**

### CODE REVIEW

**Review by JesseTNRoberts — COMMENTED** [review]
The statement of function is too long and seems to be about your device. Without clearly saying what something does, there's no way to judge if it is appropriate constrained.

File: Documentation/Signoffs/MainProcessor.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+*Figure 1: Main Processor Overview*
+
+The main processor unit is responsible for receiving, analyzing, and interpreting data from the image-detecting system to calculate and aim the interceptor. Its primary function revolves around processing data related to a golf ball's trajectory in real time. The Jetson Nano serves as the brain of the system, consolidating data from the local camera that passes the data to the main processor.
+
+## **Constraints:**
+
+| NO. |	Constraint | Origin |
+|-----|------------|--------|
+| 1	| Time Constraints - Real-time data processing for trajectory prediction for the golf ball speed varying from 1.9 seconds to 7.4 seconds | System Constraint |
+| 2	| Processing Speed - Optimization for efficient programs to calculate the speed, wire, and variable height in time, not taking longer than 40 ms per calculation iteration | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: What does this mean exactly?

File: Documentation/Signoffs/MainProcessor.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+*Figure 1: Main Processor Overview*
+
+The main processor unit is responsible for receiving, analyzing, and interpreting data from the image-detecting system to calculate and aim the interceptor. Its primary function revolves around processing data related to a golf ball's trajectory in real time. The Jetson Nano serves as the brain of the system, consolidating data from the local camera that passes the data to the main processor.
+
+## **Constraints:**
+
+| NO. |	Constraint | Origin |
+|-----|------------|--------|
+| 1	| Time Constraints - Real-time data processing for trajectory prediction for the golf ball speed varying from 1.9 seconds to 7.4 seconds | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: What does this mean?

File: Documentation/Signoffs/MainProcessor.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+| 2	| Processing Speed - Optimization for efficient programs to calculate the speed, wire, and variable height in time, not taking longer than 40 ms per calculation iteration | System Constraint |
+| 3	| Signal Interpretation Challenges - Ensuring accurate data interpretation that is received from the Arduino's needs to be 5 Volts | Programming Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: How is this a constraint?

File: Documentation/Signoffs/MainProcessor.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+| 4	| Resource Utilization - Preventing overload of system resources since the board has 1.43GHz with quad cores and 4GB RAM, the system needs to utilize them for speed efficiency but not sacrifice stability | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: How is this a constraint?

File: Documentation/Signoffs/MainProcessor.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+| 4	| Resource Utilization - Since the board has 1.43GHz with quadcores and 4GB RAM, the main processor needs to be utilized properly to prevent an overload of system resources. The system needs to use all cores and not overload the RAM for speed efficiency but not sacrifice stability | System Constraint |
+| 5 | Pausing Processes: The system needs a pause state to stop other scripts from activating firing mechanisms. | Rulebook |
+
+
+
+## **Fulfilling Constraints:**
+
+**Time Constraints:** Real-time processing of sensor data and trajectory calculations impose time constraints on the Jetson Nano. Since it has to be able to detect and calculate the proper position of the ball. Its 1.43GHz quad-core ARM Cortex-A57 the processor needs to be able to receive, process, calculate the interceptor’s path, and aim the interceptor before the golf ball gets too far down the string [3]. THe ball's travel time varies from 1.9 seconds to 7.4 seconds and the programs and data transmission needs to be optimized for an accurate and efficient system to be able to run fast enough. Delays in data acquisition, processing, or interceptor firing may affect the interception accuracy dramatically [2].
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: What is the limit? Assuming worst case scenario, how long does the processor have to do the computations? How do you know?

File: Documentation/Signoffs/MainProcessor.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+**Processing Speed:** Complex calculations and simultaneous tasks may strain the processing capabilities of the Jetson Nano, potentially leading to performance bottlenecks. Optimizing algorithms and utilizing hardware and multi-core techniques can help minimize processing speed limitations. Making sure the processor can calculate each image iteration data fast enough to give the motors enough time to get into position for firing.
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: I don't understand what is being constrained.

File: Documentation/Signoffs/MainProcessor.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+**Signal Interpretation Challenges:** Variability in wireless signal strength, interference, and environmental factors may pose challenges in accurately interpreting Arduino data. The strength of signals from Arduinos should be 5V to confirm the signal was received by the Jetson Nano.
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: Is this a constraint? I'm not sure what the means.

File: Documentation/Signoffs/MainProcessor.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+**Processing Speed:** Complex calculations and simultaneous tasks may strain the processing capabilities of the Jetson Nano, potentially leading to performance bottlenecks. Optimizing algorithms and utilizing hardware and multi-core techniques can help minimize processing speed limitations. Making sure the processor can calculate each image iteration data fast enough to give the motors enough time to get into position for firing.
+
+**Signal Interpretation Challenges:** Variability in wireless signal strength, interference, and environmental factors may pose challenges in accurately interpreting Arduino data. The strength of signals from Arduinos should be 5V to confirm the signal was received by the Jetson Nano. 
+
+**Resource Utilization:** With 4GB of LPDDR4 RAM, the Jetson Nano offers ample memory for concurrent tasks. However, efficient resource management is imperative to prevent resource overload. Optimization strategies focus on minimizing memory usage and CPU load, preserving system responsiveness [2].
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: Memory for concurrent tasks? This isn't a constraint.

File: Documentation/Signoffs/MainProcessor.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+**Pausing Processes:** Since the rulebook from Devcom states the requirement for a pause switch integrated with the interceptor to stop the machine from firing, the system will require an integrated switch to halt operations inside the Jetson Nano [9].
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: What is the constraint?

File: Documentation/Signoffs/MainProcessor.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+*Figure 1: Main Processor Overview*
+
+The main processor unit is responsible for receiving, analyzing, and interpreting data from the image-detecting system to calculate and aim the interceptor. Its primary function revolves around processing data related to a golf ball's trajectory in real time. The Jetson Nano serves as the brain of the system, consolidating data from the local camera that passes the data to the main processor.
+
+## **Constraints:**
+
+| NO. |	Constraint | Origin |
+|-----|------------|--------|
+| 1	| Time Constraints - Real-time data processing for trajectory prediction for the golf ball can take around 40 ms for each image. The main processor needs to calculate the ball data before the ball reaches the end which varies from 1.9 seconds to 7.4 seconds | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: The time that is available may vary, but you need to choose a constraint that will guarantee this system will work in every case. So, what is the appropriate constraint?

File: Documentation/Signoffs/MainProcessor.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+| 2	| Processing Speed - The main processor scripts and programs have to be optimized for efficient calculations. The scripts get the speed, wire, and variable height from input data. These calculations should not take longer than 40 ms per calculation iteration | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md]: I am not clear on why this is the constraint? I read the justification below but it did not make sense that this is the result of what is said there.

File: Documentation/Signoffs/MainProcessor.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+| 3	| Resource Utilization - Since the board has 1.43GHz with quad-cores and 4GB RAM, the main processor needs to be utilized properly to prevent an overload of system resources. The system needs to use all cores and not overload the RAM for speed efficiency but not sacrifice stability | System Constraint |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md:16]: This is not a constraint. What is the constraint here?

File: Documentation/Signoffs/MainProcessor.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+| 4 | Pausing Processes: The system needs a pause state to stop other scripts from activating firing mechanisms. | Rulebook |
```
Comment (JesseTNRoberts) [inline Documentation/Signoffs/MainProcessor.md:17]: Are there any constraints on the pause function? Or is it enough that one exists? What if I press it and it takes 3 seconds before the pause occurs?

File: Documentation/Signoffs/MainProcessor.md
Comment (micahr234) [inline Documentation/Signoffs/MainProcessor.md:1]: How are you going to communicate the position to the Arduino controlling the interceptor? Your schematic only has one GPIO connected.

What pause switch are you going to use? I see no entry on BOM.

**Review by micahr234 — COMMENTED** [review]
Need to add the pause switch to BOM

**Comment by FinlayPatoto** [comment]
> Need to add the pause switch to BOM

The Pause Switch is apart of the Extras Subsystem.

**Comment by micahr234** [comment]
Ok I will sign. You need to request a review.

**Review by micahr234 — APPROVED** [review]

**Pull request merged**